connect/eaas/workers.py

```diff
@@ -0,0 +1,244 @@
+        self.run_event = asyncio.Event()
+        self.client = AsyncConnectClient(
+            self.api_key,
+            endpoint=f'https://{self.server_address}/public/v1',
+            use_specs=False,
+        )
+        self.ws = None
+
+    async def run(self):
+        await self.run_event.wait()
+        headers = (('Authorization', self.api_key),)
+        while self.run_event.is_set():
+            try:
+                async with websockets.connect(self.get_url(), extra_headers=headers) as self.ws:
+                    while self.run_event.is_set():
+                        task_data = await self.receive()
+                        task_id = task_data['task_id']
+                        method_name = OBJ_TYPE_EXT_METHOD_MAP[task_data['object_type']]
+                        method = getattr(self.extension, method_name)
+                        request = await self.get_request(task_data)
+                        if inspect.iscoroutinefunction(method):
+                            future = asyncio.create_task(method(request))
+                        else:
+                            future = self.loop.run_in_executor(
+                                self.executor,
+                                method,
+                                request,
+                            )
+                        asyncio.create_task(self.send_result(task_id, future))
+            except websockets.exceptions.ConnectionClosedOK:
+                break
+
+    def get_url(self):  # pragma: no cover
+        raise NotImplementedError()
+
+    async def get_request(self, task_data):  # pragma: no cover
+        raise NotImplementedError()
+
+    async def build_response(self, task_id, future):  # pragma: no cover
+        raise NotImplementedError()
+
+    async def send_result(self, task_id, future):
+        await self.send(await self.build_response(task_id, future))
+
+
+class AsyncWorker(Worker):
+    async def get_request(self, task_data):
+        object_id = task_data['object_id']
+        object_type = task_data['object_type']
+        if object_type == 'asset_request':
+            return await self.client.requests[object_id].get()
+        if object_type == 'tier_config_request':
+            return await self.client('tier').config_requests[object_id].get()
+
+    def get_url(self):
+        return f'{self.base_ws_url}/tasks/async/{self.environment_id}/{self.instance_id}'
+
+    async def build_response(self, task_id, future):
+        result = None
+        try:
+            result = await future
+        except Exception:
+            return {
+                'task_id': task_id,
+                'status': 'retry',
+            }
+
+        response = {
+            'task_id': task_id,
+            'status': result.status,
+        }
+        if result.status == 'reschedule':
+            response['countdown'] = result.countdown
+        return response
+
+
+class SyncWorker(Worker):
+    async def get_request(self, task_data):
```

Comment:
I don't understand that fully here, in case of calling we get a coroutine on a function in sync class. My thinking was that base class can be sync and some successor overrides with async:
https://stackoverflow.com/a/52466625

Does that make sense?

Reply:
The name SyncWorker doesn't mean that is synchronous. In fact we have decided to change its name to InteractiveTaskWorker. We mean the worker that handles tasks like real time validation.